lib/domain/provider/provider_model_async_result.dart

```diff
@@ -0,0 +1,21 @@
+import 'dart:ui';
+
+import 'package:async/async.dart';
+
+class ProviderModelAsyncResult<T, Y extends Exception> {
+  Future<void> set(
+      Future<Result<T>> futureResult, VoidCallback isLoadingChanged) async {
+    isLoading = true;
+    isLoadingChanged();
+
+    result = await futureResult;
+
+    isLoading = false;
+    isLoadingChanged();
+  }
+
+  bool get hasData => isLoading != null || result != null;
```

Comment:
Aby być gotowym na NNBD myślę że możemy na potrzeby tej klasy założyć, że isLoading nigdy nie jest nullem. Można zainicjować na false w konstruktorze, i:
```suggestion
  bool get hasData => !isLoading || result != null;
```